lib/queue.js

```diff
@@ -830,16 +830,32 @@ class Queue extends Emitter {
       args[i] = arguments[i];
     }
 
-    return this._commandable().then((client) => {
-      // Get the sha for the script after we're ready to avoid a race condition
-      // with the lua script loader.
-      args[0] = lua.shas[name];
-
+    function _evalScriptInner(client) {
       const promise = helpers.deferred();
       args.push(promise.defer());
       client.evalsha.apply(client, args);
       return promise;
-    });
+    }
+
+    return this._commandable()
+      .then((client) => {
+        // Get the sha for the script after we're ready to avoid a race
+        // condition with the lua script loader.
+        args[0] = lua.shas[name];
+        return _evalScriptInner(client);
+      })
+      .catch((err) => {
+        // In cases where the Redis server fails over, the script cache is
+        // cleared (even when the cache itself is still intact.) We detect this
+        // condition and reload the scripts.
+        if (err.code !== 'NOSCRIPT') {
+          throw err;
+        }
+        return lua.buildCache(this.client).then((client) => {
```

Comment:
The promise produced by `buildCache` will not resolve to the client. We'll want to use `_commandable` again anyway in case the `Queue` is in the process of exiting.

Reply:
d'oh!